src/components/vm/disks/diskAdd.jsx

```diff
@@ -365,511 +359,498 @@ const CustomPath = ({ idPrefix, onValueChanged, device, validationFailed, hideDe
-        const availableTarget = getNextAvailableTarget(existingTargets, defaultBus);
-        const sortFunction = (poolA, poolB) => poolA.name.localeCompare(poolB.name);
-        let defaultPool;
-        if (storagePools.length > 0)
-            defaultPool = storagePools
-                    .map(pool => ({ name: pool.name, type: pool.type }))
-                    .sort(sortFunction)[0];
-
-        return {
-            storagePools,
-            vm,
-            vms,
-            file: "",
-            device: "disk",
-            storagePoolName: defaultPool && defaultPool.name,
-            storagePoolType: defaultPool && defaultPool.type,
-            mode: isMediaInsertion ? CUSTOM_PATH : CREATE_NEW,
-            volumeName: "",
-            existingVolumeName: undefined,
-            size: 1,
-            unit: units.GiB.name,
-            format: defaultPool && getDefaultVolumeFormat(defaultPool),
-            target: availableTarget,
-            permanent: !domainIsRunning(vm.state), // default true for a down VM; for a running domain, the disk is attached tentatively only
-            hotplug: domainIsRunning(vm.state), // must be kept false for a down VM; the value is not being changed by user
-            addDiskInProgress: false,
-            cacheMode: 'default',
-            busType: defaultBus,
-            updateDisks: false,
-            serial: "",
-        };
-    }
+    const Dialogs = useDialogs();
+
+    const defaultPool = useMemo(() => storagePools?.[0], [storagePools]);
 
-    componentDidMount() {
+    // VM params that should be refreshed when VM, storage Pools or mode changes
+    const initialDiskParams = useMemo(() => ({
+        connectionName: vm.connectionName,
+        existingVolumeName: defaultPool && getDefaultVolumeName(defaultPool, vm),
+        storagePool: defaultPool,
+        type: mode === CUSTOM_PATH ? "file" : "volume",
+        permanent: !domainIsRunning(vm.state)
+    }), [defaultPool, mode, vm]);
+    const storagePoolName = diskParams.storagePool?.name;
+
+    useEffect(() => {
         // Refresh storage volume list before displaying the dialog.
         // There are recently no Libvirt events for storage volumes and polling is ugly.
         // https://bugzilla.redhat.com/show_bug.cgi?id=1578836
-        storagePoolGetAll({ connectionName: this.props.vm.connectionName })
-                .then(() => this.setState({ dialogLoading: false, ...this.initialState }))
-                .catch(exc => this.dialogErrorSet(_("Storage pools could not be fetched"), exc.message));
-    }
-
-    validateParams() {
-        const validationFailed = {};
+        storagePoolGetAll({ connectionName: vm.connectionName })
+                .always(() => {
+                    setStoragePools(getVmStoragePools(vm.connectionName).sort(sortFunction));
+                })
+                .catch(exc => dialogErrorSet(_("Storage pools could not be fetched"), exc.message));
```

Comment:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1303-20231120-122732-13cb53e1-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)